### DISCUSSION TITLE
OpenSSL 3.0.9 unable to find config file

### DISCUSSION THREAD

**Discussion by sabadjev** [openssl/openssl#25258] [Q&A] [answered]
I am running into issues with loading the FIPS provider (OSSL_PROVIDER_load(NULL, "fips") returns NULL). Specifically, judging by the error stack, it looks like it's unable to find my config file (unsure of whether it's unable to find openssl.cnf or just fipsmodule.cnf):

B05D0000:error:1C8000D5:Provider routines:SELF_TEST_post:missing config data:providers\fips\self_test.c:290:
B05D0000:error:1C8000E0:Provider routines:ossl_set_error_state:fips module entering error state:providers\fips\self_test.c:388:
B05D0000:error:1C8000D8:Provider routines:OSSL_provider_init_int:self test post failure:providers\fips\fipsprov.c:707:
B05D0000:error:078C0105:common libcrypto routines:provider_init:init fail:crypto\provider_core.c:932:name=fips

I have tried editing OPENSSL_CONF as well as my config files but it's ending up with the exact same result. All of the advice I've found so far is to try editing OPENSSL_CONF or the config files. Is there any insight into why this might be happening and how to fix it?

More information:
I am using a Windows x64 build of OpenSSL. The providers (both FIPS and default) are from version 3.0.9 but openssl version -v returns OpenSSL 3.3.0. OPENSSLDIR is a directory in the Z drive for some reason. I have added both a system and user level environmental variable called OPENSSL_CONF and set them to the path to my openssl.cnf file. My config files are below:

openssl.cnf:
config_diagnostics = 1
openssl_conf = openssl_init

.include (path to fips config)\fipsmodule.cnf

[openssl_init]
providers = provider_sect
alg_section = algorithm_sect

[provider_sect]
fips = fips_sect
base = base_sect

[base_sect]
activate = 1

[algorithm_sect]
default_properties = fips=yes

fipsmodule.cnf:
[fips_sect]
install-version = 1
conditional-errors = 1
security-checks = 1
tls1-prf-ems-check = 0
drbg-no-trunc-md = 0
module-mac = (MAC)

EDIT: I have also just realized that fips.dll is still loading from the default install location (C:\Program Files\OpenSSL\lib\ossl-modules\fips.dll), not the location inside my app's files that I want it to use. Maybe this is relevant?

**Comment by kulkarniamit** [comment, marked as answer]
> unsure of whether it's unable to find openssl.cnf or just fipsmodule.cnf

It may be helpful to start with [process monitor](https://learn.microsoft.com/en-us/sysinternals/downloads/procmon) and verify the path of `openssl.cnf`, `fips.dll` and `fipsmodule.cnf` being used while running your program.

> EDIT: I have also just realized that fips.dll is still loading from the default install location (C:\Program Files\OpenSSL\lib\ossl-modules\fips.dll), not the location inside my app's files that I want it to use. Maybe this is relevant?

The `module-mac` in your `fipsmodule.cnf` is based on the `fips.dll` that was used during `openssl fipsinstall ...` command. Your application should be using the same `fips.dll` that was used while generating `fipsmodule.cnf`. Otherwise, the provider fails to load. You can use `OPENSSL_MODULES` [env variable](https://docs.openssl.org/master/man7/openssl-env/) to set the correct path of `fips.dll` for your application. 
To ensure control over paths, I would use:
* `OPENSSL_CONF` pointing to full path of `openssl.cnf` that has the path to `fipsmodule.cnf`
* `OPENSSL_MODULES` pointing to full path of directory where `fips.dll` is located (Example: `C:\tmp\lib\ossl-modules`) 

I'd also recommend starting with verifying if your openssl binary is able to load fips provider from the correct location using `openssl list -providers` command as shown here: https://github.com/openssl/openssl/discussions/25173#discussioncomment-10389365

Additionally try the instructions from these similar questions:
* https://github.com/openssl/openssl/discussions/21602
* https://github.com/openssl/openssl/discussions/21447
* https://github.com/openssl/openssl/discussions/25173#discussioncomment-10329147

**Reply by sabadjev** [reply]
Thank you so much for the reply! I was able to narrow my issues down quite a bit thanks to this. Now I am facing issues with fips.dll. When I run my app through VS Debugger, it finds the correct fips.dll (verified through Process Monitor). However, when I double-click on a file to open it using the app, it looks for fips.dll in another location where it doesn't find it and error. If I place a fips.dll in that other location, it instead throws the same error stack as the original post (but the way I see it the real issue is that it's looking in the wrong place, not that fips.dll is missing from the erroneous path).

**Reply by sabadjev** [reply]
I figured out the issue, this time my environmental variables were pointed at the wrong place.
